include/boost/decimal/detail/cmath/ellint_1.hpp

```diff
@@ -29,124 +30,133 @@ namespace decimal {
 
-    if(phi > 1 / std::numeric_limits<T>::epsilon())
-    {
-        // Phi is so large that phi%pi is necessarily zero (or garbage),
-        // just return the second part of the duplication formula:
-        result = 2 * phi * ellint_k_imp(k) / numbers::pi_v<T>;
+      const T top { phi * (-60 + (-12 + 17 * m2) * phi_sq) };
+      const T bot { -60 + 3 * (-4 + 9 * m2) * phi_sq };
+
+      result = top / bot;
     }
     else
     {
-        // Carlson's algorithm works only for |phi| <= pi/2,
-        // use the integrand's periodicity to normalize phi
-        //
-        // Xiaogang's original code used a cast to long long here
-        // but that fails if T has more digits than a long long,
-        // so rewritten to use fmod instead:
-        //
-        T rphi = fmod(phi, half_pi);
-        T m = round((phi - rphi) / half_pi);
-        int s = 1;
-
-        if (fmod(m, T{2}) > T{5, -1})
-        {
-            m += 1;
-            s = -1;
-            rphi = half_pi - rphi;
-        }
-
-        T sinp = sin(rphi);
-        sinp *= sinp;
-
-        if (sinp * k * k >= 1)
-        {
-            return std::numeric_limits<T>::signaling_NaN();
-        }
-
-        T cosp = cos(rphi);
-        cosp *= cosp;
-        if (sinp > (std::numeric_limits<T>::min)())
-        {
-            BOOST_DECIMAL_ASSERT(rphi != 0); // precondition, can't be true if sin(rphi) != 0.
-            //
-            // Use http://dlmf.nist.gov/19.25#E5, note that
-            // c-1 simplifies to cot^2(rphi) which avoid cancellation:
-            //
-            T c = 1 / sinp;
-            result = static_cast<T>(s * ellint_impl::ellint_rf_imp(cosp / sinp, c - k * k, c));
-        }
-        else
-        {
-            result = s * sin(rphi);
-        }
-
-        if(m != 0)
-        {
-            result += m * ellint_k_imp(k);
-        }
+      constexpr T my_pi_half { numbers::pi_v<T> / 2 };
+
+      T k_pi       = static_cast<int>(phi / numbers::pi_v<T>);
+      T phi_scaled = phi - (k_pi * numbers::pi_v<T>);
+
+      const bool b_neg { phi_scaled > my_pi_half };
+
+      if(b_neg)
+      {
+        ++k_pi;
+
+        phi_scaled = -(phi_scaled - numbers::pi_v<T>);
+      }
+
+      T Km { };
+
+      detail::ellint_detail::elliptic_series::agm(phi_scaled, m, result, Km);
+
+      if(b_neg)
+      {
+        result = -result;
+      }
+
+      result += ((k_pi * Km) * 2);
     }
+  }
 
-    return invert ? -result : result;
+  return result;
 }
 
 template <typename T>
-constexpr auto comp_ellint_1_impl(T k) noexcept
+constexpr auto comp_ellint_1_impl(T m) noexcept
     BOOST_DECIMAL_REQUIRES(detail::is_decimal_floating_point_v, T)
 {
-   return detail::ellint_k_imp(k);
+  constexpr T one  { 1 };
+
+  T result { };
+
+  const auto fpc_m   = fpclassify(m);
+
+  if(fpc_m == FP_ZERO)
+  {
+    result = numbers::pi_v<T> / 2;
+  }
+  else if((fabs(m) > one) || (fpc_m != FP_NORMAL))
+  {
+    result = std::numeric_limits<T>::quiet_NaN();
```

Comment:
Should we use qNaN or sNaN in the cmath functions in this case? I think you have been using qNaN and I have used a few sNaNs. The C++ standard just says NAN

Reply:
Good question. I was hoping to use exclusively one or the other in the project. Of course, on bare metal signalling NaN would be difficult. So if you do intend to have any small-weight clients, I'd say use `quiet_NaN()` exclusively. But I won't make a big fuss about it either way.

Reply:
I'll open an issue to covert to `quiet_NaN()`. Since `issignaling` isn't standard theres not even a good way to differentiate at the moment. Might as well use the one we know is good everywhere.